subprojects/core/src/main/java/org/gradle/process/internal/JvmOptions.java

```diff
@@ -135,7 +136,10 @@ protected void formatSystemProperties(Map<String, ?> properties, List<String> ar
         if (assertionsEnabled) {
             args.add("-ea");
         }
-        if (debug) {
+
+        if (debugOptions != null) {
+            args.add(debugOptions);
+        } else if (debug) {
```

Comment:
setDebugOptions should also set `debug` to `true`, as e.g. the `Test` task depends on that flag